orttraining/orttraining/models/runner/training_runner.cc

```diff
@@ -201,45 +277,310 @@ Status TrainingRunner::Run(IDataLoader* training_data_loader, IDataLoader* test_
+    // Set up tensor to be fetched when doing model update. 
+
+    if (params_.pipeline_parallel_size > 1) {
+      // If pipeline is used, we need to filter out fetches which are not in this pipeline stage.
+
+      for (size_t i = 0; i < params_.fetch_names.size(); ++i) {
+        const auto name = params_.fetch_names[i];
+        auto it = std::find(allowed_fetch_names.begin(), allowed_fetch_names.end(), name);
+        if (it == allowed_fetch_names.end()) {
+          continue;
+        }
+        fetch_names.push_back(name);
+      }
+    } else {
+      // No pipeline. All fetched names should appear in the graph handled by this process.
+      fetch_names = params_.fetch_names;
+    }
+
+    if (params_.use_mixed_precision) {
+      auto it = opt_graph_outputs_.find(OptimizerOutputKey::GradientAllIsFinite);
+      ORT_RETURN_IF(it == opt_graph_outputs_.end(), "Gradient norm's IsFinite output is missing in the optimizer output");
       fetch_names.push_back(it->second);
+      if (params_.use_adasum) {
+        it = opt_graph_outputs_.find(OptimizerOutputKey::DeltaAllIsFinite);
+        ORT_RETURN_IF(it == opt_graph_outputs_.end(), "Adasum delta's IsFinite output is missing in the optimizer output");
+        fetch_names.push_back(it->second);
+      }
+    }
+  } else if (mode == GradientAccumulateStep) {
+    // Set up tensor to be fetched when doing gradient accumulation. 
+
+    if (params_.gradient_accumulation_steps > 1) {
+      auto it = opt_graph_outputs_.find(OptimizerOutputKey::GradientAccumulation);
+      ORT_RETURN_IF(it == opt_graph_outputs_.end(), "Gradient accumulation output is missing in the optimizer output");
+      fetch_names.push_back(it->second);
+    }
+
+    // Always execute event operators to avoid deadlock if pipeline is used.
+    // TODO: create a list of must-to-fetch tensors and pass it to all graph transformer.
+    if (params_.pipeline_parallel_size) {
+        if (!pipeline_context_.forward_waited_output_name.empty()) {
+          fetch_names.push_back(pipeline_context_.forward_waited_output_name);
+        }
+        if (!pipeline_context_.forward_recorded_output_name.empty()) {
+          fetch_names.push_back(pipeline_context_.forward_recorded_output_name);
+        }
+        if (!pipeline_context_.backward_waited_output_name.empty()) {
+          fetch_names.push_back(pipeline_context_.backward_waited_output_name);
+        }
+        if (!pipeline_context_.backward_recorded_output_name.empty()) {
+          fetch_names.push_back(pipeline_context_.backward_recorded_output_name);
+        }
+    }
+  } else if (mode == EvaluateStep) {
+    // Set up tensor to be fetched when doing model evaluation.
+    // Ideally, this path should not fetch optimizer and gradient accumulation. 
+    // This path may fetch predicted scores, loss value, and so on.
+
+    if (params_.pipeline_parallel_size > 1) {
+      // If pipeline is used, we need to filter out fetches which are not in this pipeline stage.
```

Comment:
>// If pipeline is used, we need to filter out fetches which are not in this pipeline stage. [](start = 6, length = 91)

are there cases outside of pipeline, that we might need to filter out fetches?

Reply:
No. Only pipeline parallelism can tolerate this -- user (from main.cc) should only fetch whatever exists in the model.